torch/csrc/api/include/torch/data/samplers/distributed.h

```diff
@@ -0,0 +1,113 @@
+#pragma once
+
+#include <torch/csrc/WindowsTorchApiMacro.h>
+#include <torch/data/samplers/base.h>
+
+#include <cstddef>
+#include <vector>
+
+namespace torch {
+namespace serialize {
+class OutputArchive;
+class InputArchive;
+} // namespace serialize
+} // namespace torch
+
+namespace torch {
+namespace data {
+namespace samplers {
+
+/// A `Sampler` that selects a subset of indices to sample from and defines a
+/// sampling behavior. In a distributed setting, this selects a subset of the
+/// indices depending on the provided num_replicas and rank parameters. The
+/// `Sampler` performs a rounding operation based on the `allow_duplicates`
+/// parameter to decide the local sample count.
+class DistributedSampler : public Sampler<> {
```

Comment:
This is kind of a needless specialization of `Sampler`. Why can't I have a `DistributedSampler` of another `BatchRequest` type? What if I don't want to get back `vector<size_t>`? It seems odd that `Sampler` be general but `DistributedSampler` not

Reply:
Agreed, we should preserve the possibility of specifying a different request type.

Reply:
make sense, added the template type to DistributedSampler.